tests/nonbonded/test_nonbonded_mol_energy.py

```diff
@@ -0,0 +1,227 @@
+    bond_pot = next(bp for bp in bps if isinstance(bp.potential, HarmonicBond)).potential
+
+    all_group_idxs = get_group_indices(get_bond_list(bond_pot), conf.shape[0])
+
+    assert len(all_group_idxs) >= num_mols
+
+    # Get the indices of atoms within each molecule
+    group_idxs = all_group_idxs[:num_mols]
+
+    conf_idxs = np.array(group_idxs).reshape(-1)
+
+    conf = conf[conf_idxs]
+    # Compute the box
+    box_lengths = np.amax(conf, axis=0) - np.amin(conf, axis=0)
+    box = np.eye(3, dtype=np.float64) * box_lengths
+
+    N = conf.shape[0]
+
+    params = nb.params[conf_idxs]
+
+    beta = nb.potential.beta
+    cutoff = nb.potential.cutoff
+    klass = custom_ops.NonbondedMolEnergyPotential_f32
+    if precision == np.float64:
+        klass = custom_ops.NonbondedMolEnergyPotential_f64
+
+    mover = BDExchangeMove(beta, cutoff, params, group_idxs, DEFAULT_TEMP)
+
+    mol_by_mol_pot = klass(N, group_idxs, beta, cutoff)
+
+    def u_ref(x, box, params):
+        return mover.batch_U_fn(x, box, mover.all_a_idxs, mover.all_b_idxs)
+
+    def u_test(x, box, params):
+        mol_energies = mol_by_mol_pot.execute(x, params, box)
+        assert mol_energies.shape == (len(group_idxs),)
+
+        # Make sure running again gives bitwise identical results
+        comp_mol_energies = mol_by_mol_pot.execute(x, params, box)
+        np.testing.assert_array_equal(mol_energies, comp_mol_energies)
+        return mol_energies
+
+    np.testing.assert_allclose(u_test(conf, box, params), u_ref(conf, box, params), rtol=rtol, atol=atol)
+
+    mols_to_move = rng.choice(np.arange(len(group_idxs)), size=moves)
+
+    # empirical threshold based on testing
+    threshold = 1e8
+
+    for mol_idx in mols_to_move:
+        translation = np.diag(box) * rng.uniform(size=3)
+        atom_idxs = np.array(group_idxs[mol_idx])
+        moved_coords = randomly_rotate_and_translate(conf[atom_idxs], translation)
+        updated_conf = conf.copy()
+        updated_conf[atom_idxs] = moved_coords
+        test_mol_energies = u_test(updated_conf, box, params)
+        ref_mol_energies = u_ref(updated_conf, box, params)
+        large_energy_indices = np.argwhere(np.abs(ref_mol_energies) >= threshold)
+        comparable_energies = np.delete(np.arange(len(test_mol_energies)), large_energy_indices)
+        np.testing.assert_allclose(
+            ref_mol_energies[comparable_energies], test_mol_energies[comparable_energies], rtol=rtol, atol=atol
+        )
+        # Pull out nans, as they are effectively greater than the threshold
+        non_nan_idx = np.isfinite(test_mol_energies[large_energy_indices])
+        # Large energies are not reliable, so beyond the threshold we simply verify that both the reference and test both exceed the threshold
+        assert np.all(np.abs(test_mol_energies[large_energy_indices][non_nan_idx]) >= threshold)
+
+
+@pytest.mark.parametrize("precision,atol,rtol", [(np.float64, 1e-8, 1e-8), (np.float32, 5e-4, 3e-3)])
```

Comment:
Had to tweak float32 tolerances to be pretty low, not clear to me how to rigorously approach this. Since the float64 is working across 1e-8 tolerances I felt comfortable lowering the float32 tolerances.

Reply:
Hm  this is still a lot lower than we typically allow for. Typically, we compute *total* energies (as a scalar), however, here we're computing *per molecule* energies (as a vector). So the effect seems even bigger than desired. I do think we should investigate this a bit more:

Can you print out each component of the energy vector and see if the discrepancy is coming from waters that have *large* molecular energies or waters that have *small* molecular energies? If the discrepancy is coming from waters that have *small* molecular energies, this is less worrisome (since they will have negligible weights after exponentiation), but if it's coming from waters that have *large* molecular energies, then this is definitely a problem.

Reply:
Need to investigate in more detail, but from what I remember the absolute errors were about 2.0 and the relative error was 0.002, so would indicate a large energy disagreeing. And we already have issues with large (magnitude) molecular energies disagreeing due to the int128 handling of overflows.

Reply:
This looks like a system straight from build_water_system, there might be a clash?

Reply:
Will try minimizing and see if that helps. Though most of the time this is going to be looking at clashy systems, so not clear how that makes a difference?

Reply:
It's the expectations compared to known systems we've benchmarked and tested before (which IIRC are non-clashy systems).

1. For a non-clashy system, we do *not* expect to have large errors in 32bit (I was mainly assuming this was the case, and hence why I requested explicitly a clashy test system).
2. For a clashy system, we expect that the *clashy* atoms/molecules have a large discrepancy in 32bit.

Reply:
Thanks for clarify. You were spot on, the system contained a clash and because it hadn't been minimized the difference was greater. Added 1000 steps of 1.5e-3 timestep (2.5e-3 blows up) and was able to decrease the tolerances: https://github.com/proteneer/timemachine/pull/1198/commits/c518a0c1da4dae3f5f6522a4781b8937486b579d